src/main/scala/com/fulcrumgenomics/bam/CallOverlappingConsensusBases.scala

```diff
@@ -0,0 +1,182 @@
+  @arg(doc="The number of threads to use while consensus calling.") val threads: Int = 1,
+  @arg(flag='S', doc="The sort order of the output. If not given, output will be in the same order as input if the input.")
+  val sortOrder: Option[SamOrder] = None,
+  @arg(doc="""If the read and mate bases disagree at a given reference position, true to mask (make 'N') the read and mate
+             |bases, otherwise pick the base with the highest base quality and return a base quality that's the difference
+             |between the higher and lower base qualities.""")
+  val onlyMaskDisagreements: Boolean = false,
+  @arg(doc= """If the read and mate bases agree at a given reference position, true to for the resulting base quality
+              |to be the maximum base quality, otherwise the sum of the base qualities.""")
+  val maxQualOnAgreement: Boolean = false
+) extends FgBioTool with LazyLogging {
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  private case class ThreadData(caller: OverlappingBasesConsensusCaller, templateMetric: CallOverlappingConsensusBasesMetric, basesMetric: CallOverlappingConsensusBasesMetric)
+  private object ThreadData {
+    def apply(): ThreadData = ThreadData(
+      caller         = new OverlappingBasesConsensusCaller(onlyMaskDisagreements=onlyMaskDisagreements, maxQualOnAgreement=maxQualOnAgreement),
+      templateMetric = CallOverlappingConsensusBasesMetric(tpe=CountType.Templates),
+      basesMetric    = CallOverlappingConsensusBasesMetric(tpe=CountType.Bases)
+    )
+  }
+
+  override def execute(): Unit = {
+    val source           = SamSource(input)
+    val writer           = SamWriter(output, source.header)
+    val progress         = new ProgressLogger(logger)
+    val templateIterator = Bams.templateIterator(source)
+    val threadData       = new IterableThreadLocal(() => ThreadData())
+
+    ParIterator(templateIterator, threads=threads)
+      .map { template =>
+        val threadDatum = threadData.get()
+        threadDatum.synchronized {
+          // update metrics
+          threadDatum.templateMetric.total += 1
+          threadDatum.basesMetric.total += template.primaryReads.map(_.length).sum
+          // corrects
+          val stats = threadDatum.caller.call(template)
+          val overlappingBases = stats.r1OverlappingBases + stats.r2OverlappingBases
+          val correctedBases   = stats.r1CorrectedBases + stats.r2CorrectedBases
+          if (overlappingBases > 0) {
+            threadDatum.templateMetric.overlapping += 1
+            threadDatum.basesMetric.overlapping += overlappingBases
+            if (correctedBases > 0) {
+              threadDatum.templateMetric.corrected += 1
+              threadDatum.basesMetric.corrected += correctedBases
+            }
+          }
+        }
+        template.allReads.foreach(progress.record)
+        template
+      }.toAsync(ParIterator.DefaultChunkSize * 8).foreach { template =>
+      writer ++= template.allReads
```

Comment:
I'm realizing that doing this invalidates the NM and MD tags on reads that we edit.  Should we remove those tags?  Recalculate?  Just document this fact?

Reply:
Recalculate isn't too hard these days.  I'l fix it.